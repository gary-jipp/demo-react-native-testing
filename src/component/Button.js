import React from 'react';
import {StyleSheet, Text, View, Pressable, } from 'react-native';

const Button = (props) => {
  // console.log(`Button.${props.title}.disabled:`, props.disabled);

  return (
    <Pressable
      testID="button"
      accessibilityRole="button"
      disabled={props.disabled}
      style={props.disabled || styles.pressable}
      onPress={props.onPress} >
      <View style={[styles.button, props.style, props.disabled && styles.disabled]}>
        <Text style={[styles.text, props.textStyle]}>{props.title}</Text>
      </View>
    </Pressable >
  );
};

const styles = StyleSheet.create({
  pressable: function({pressed}) {
    return {opacity: pressed ? 0.8 : 1.0};
  },
  button: {
    alignItems: 'center',
    justifyContent: 'center',
    paddingVertical: 12,
    paddingHorizontal: 20,
    borderRadius: 5,
    backgroundColor: '#2196F3',
    elevation: 0,
  },
  text: {
    fontSize: 16,
    lineHeight: 21,
    fontWeight: 'bold',
    letterSpacing: 0.25,
    color: '#EEE',
  },
  disabled: {
    backgroundColor: '#EEE',
  }
});

export default Button;