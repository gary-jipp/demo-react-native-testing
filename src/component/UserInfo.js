import React from "react";
import {StyleSheet, Text, View} from "react-native";
import Slogan from "./Slogan";

export default UserInfo = (props) => {
  return (
    <View style={styles.container} accessibilityLabel="Info" >
      <UserItem accessibilityLabel="User Name" label="Name" value={props.user.name} />
      <UserItem accessibilityLabel="City" label="City" value={props.user.address.city} />
      <UserItem label="Company" value={props.user.company.name} />
      <UserItem label="Phone" value={props.user.phone} />
      <UserItem label="Website" value={props.user.website} />
      <Slogan text={props.user.company.bs} />
    </View>
  );
};

const UserItem = ({label, value, accessibilityLabel}) => (
  <View accessibilityLabel="User Item"
    style={styles.userItem} >
    <View style={styles.itemLabel}>
      <Text style={styles.text}>{label}</Text>
    </View>
    <View style={styles.itemValue}>
      <Text
        accessibilityLabel={accessibilityLabel}
        style={styles.text}>{value}</Text>
    </View>
  </View >
);

const styles = StyleSheet.create({
  container: {
    backgroundColor: "#fff",
    paddingHorizontal: 0,
    // borderWidth: 1,
    // borderColor: 'red',
  },
  userItem: {
    flexDirection: "row",
    marginBottom: 10,
    marginTop: 10,
    // borderWidth: 1,
    // borderColor: 'red',
  },
  itemLabel: {
    marginRight: 2,
    width: '35%',
    // borderWidth: 1,
    // borderColor: 'red',
  },
  itemValue: {
    // borderWidth: 1,
    // borderColor: 'red',
  },
  text: {
    color: "#353031",
    fontWeight: "500",
    fontSize: 19,
    // borderWidth: 1,
    // borderColor: 'red',
  },
});