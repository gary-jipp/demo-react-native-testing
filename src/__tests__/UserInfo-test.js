import React from 'react';
import {render, cleanup, getQueriesForElement, fireEvent} from "@testing-library/react-native";
import UserInfo from 'component/UserInfo';
import Button from 'component/Button';

// unmounts the components after each test
afterEach(cleanup);

describe("Testing Library Tests for UserInfo", () => {

  it("Renders a User", () => {
    const component = <UserInfo user={mockUser} />;
    const {container, getByRole, getAllByLabelText, getByLabelText, getByText, toJSON} = render(component);
    expect(toJSON()).toMatchSnapshot();

    let info = getByLabelText("Info");
    // Exposes the underlying test-renderer methods
    const button = info.findByType(Button);
    // console.log(button.props);

    let queries = getQueriesForElement(info);
    let text = queries.getByText("Leanne Graham");  // Self-asserting

    // Notice arrays are not search  result objects
    const items = getAllByLabelText("Info");
    // console.log(items.length);
    // queries = getQueriesForElement(items);  // bad result.  Arrays are just arrays
    queries = getQueriesForElement(items[0]);  // Now it works.  Arrays are just arrays
    text = queries.findByText("Leanne Graham");
    expect(text).not.toBeNull();    // Really, a snapshot does this

    const sloganButton = getByRole("button");
    console.log(sloganButton.props);
    fireEvent.press(sloganButton);    // Can't presss something that's not pressable


    // Self asserting
    const bs = getByText('harness real-time e-markets');
    // console.log(bs);

    // findBy does not assert, always returns "something", even null

    // console.log(items.length);
    // console.log(items[6].children);

    // console.log(items);
    // const name = getByLabelText("Name");

  });

});

const mockUser =
{
  "id": 1,
  "name": "Leanne Graham",
  "username": "Bret",
  "email": "Sincere@april.biz",
  "address": {
    "street": "Kulas Light",
    "suite": "Apt. 556",
    "city": "Gwenborough",
    "zipcode": "92998-3874",
    "geo": {
      "lat": "-37.3159",
      "lng": "81.1496"
    }
  },
  "phone": "1-770-736-8031 x56442",
  "website": "hildegard.org",
  "company": {
    "name": "Romaguera-Crona",
    "catchPhrase": "Multi-layered client-server neural-net",
    "bs": "harness real-time e-markets"
  }
};