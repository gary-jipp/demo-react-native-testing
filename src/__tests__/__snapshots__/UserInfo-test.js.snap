// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Library Tests for UserInfo Renders a User 1`] = `
<View
  accessibilityLabel="Info"
  style={
    Object {
      "backgroundColor": "#fff",
      "paddingHorizontal": 0,
    }
  }
>
  <View
    accessibilityLabel="User Item"
    style={
      Object {
        "flexDirection": "row",
        "marginBottom": 10,
        "marginTop": 10,
      }
    }
  >
    <View
      style={
        Object {
          "marginRight": 2,
          "width": "35%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#353031",
            "fontSize": 19,
            "fontWeight": "500",
          }
        }
      >
        Name
      </Text>
    </View>
    <View
      style={Object {}}
    >
      <Text
        accessibilityLabel="User Name"
        style={
          Object {
            "color": "#353031",
            "fontSize": 19,
            "fontWeight": "500",
          }
        }
      >
        Leanne Graham
      </Text>
    </View>
  </View>
  <View
    accessibilityLabel="User Item"
    style={
      Object {
        "flexDirection": "row",
        "marginBottom": 10,
        "marginTop": 10,
      }
    }
  >
    <View
      style={
        Object {
          "marginRight": 2,
          "width": "35%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#353031",
            "fontSize": 19,
            "fontWeight": "500",
          }
        }
      >
        City
      </Text>
    </View>
    <View
      style={Object {}}
    >
      <Text
        accessibilityLabel="City"
        style={
          Object {
            "color": "#353031",
            "fontSize": 19,
            "fontWeight": "500",
          }
        }
      >
        Gwenborough
      </Text>
    </View>
  </View>
  <View
    accessibilityLabel="User Item"
    style={
      Object {
        "flexDirection": "row",
        "marginBottom": 10,
        "marginTop": 10,
      }
    }
  >
    <View
      style={
        Object {
          "marginRight": 2,
          "width": "35%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#353031",
            "fontSize": 19,
            "fontWeight": "500",
          }
        }
      >
        Company
      </Text>
    </View>
    <View
      style={Object {}}
    >
      <Text
        style={
          Object {
            "color": "#353031",
            "fontSize": 19,
            "fontWeight": "500",
          }
        }
      >
        Romaguera-Crona
      </Text>
    </View>
  </View>
  <View
    accessibilityLabel="User Item"
    style={
      Object {
        "flexDirection": "row",
        "marginBottom": 10,
        "marginTop": 10,
      }
    }
  >
    <View
      style={
        Object {
          "marginRight": 2,
          "width": "35%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#353031",
            "fontSize": 19,
            "fontWeight": "500",
          }
        }
      >
        Phone
      </Text>
    </View>
    <View
      style={Object {}}
    >
      <Text
        style={
          Object {
            "color": "#353031",
            "fontSize": 19,
            "fontWeight": "500",
          }
        }
      >
        1-770-736-8031 x56442
      </Text>
    </View>
  </View>
  <View
    accessibilityLabel="User Item"
    style={
      Object {
        "flexDirection": "row",
        "marginBottom": 10,
        "marginTop": 10,
      }
    }
  >
    <View
      style={
        Object {
          "marginRight": 2,
          "width": "35%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#353031",
            "fontSize": 19,
            "fontWeight": "500",
          }
        }
      >
        Website
      </Text>
    </View>
    <View
      style={Object {}}
    >
      <Text
        style={
          Object {
            "color": "#353031",
            "fontSize": 19,
            "fontWeight": "500",
          }
        }
      >
        hildegard.org
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "justifyContent": "center",
        "paddingHorizontal": 0,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="button"
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#2196F3",
              "borderRadius": 5,
              "elevation": 0,
              "justifyContent": "center",
              "paddingHorizontal": 20,
              "paddingVertical": 12,
            },
            undefined,
            false,
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#EEE",
                "fontSize": 16,
                "fontWeight": "bold",
                "letterSpacing": 0.25,
                "lineHeight": 21,
              },
              undefined,
            ]
          }
        >
          Slogan
        </Text>
      </View>
    </View>
    <Text
      style={
        Object {
          "color": "#353031",
          "fontSize": 18,
          "fontWeight": "500",
        }
      }
      testID="slogan"
    >
      
    </Text>
  </View>
</View>
`;
